File: src/datatypes/device.ts
import { BaseEntity, DeviceType } from './types';

export class Device implements BaseEntity {
    device_id!: string;
    user_id!: string;
    device_type!: DeviceType;
    os_type!: string;
    os_version!: string;
    app_version!: string;
    available_sensors!: string[];
    capabilities!: Record<string, any>;
    created_at!: Date;
    last_seen!: Date;
    updated_at!: Date;

    constructor(data: Partial<Device>) {
        Object.assign(this, data);
    }
}

export interface DeviceCapabilities {
    hasCamera: boolean;
    hasMicrophone: boolean;
    hasGPS: boolean;
    hasAccelerometer: boolean;
    hasGyroscope: boolean;
    hasMagnetometer: boolean;
    hasProximity: boolean;
    hasLight: boolean;
    hasPressure: boolean;
    hasTemperature: boolean;
    hasHumidity: boolean;
    hasStepCounter: boolean;
    hasHeartRate: boolean;
    hasECG: boolean;
    hasBloodOxygen: boolean;
    hasStress: boolean;
    hasCompass: boolean;
    screenDetails: {
        width: number;
        height: number;
        density: number;
        refreshRate: number;
    };
} 

File: src-tauri/src/datatypes/device.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use crate::datatypes::types::DeviceType;

#[derive(Debug, Serialize, Deserialize)]
pub struct Device {
    pub device_id: String,
    pub user_id: String,
    pub device_type: DeviceType,
    pub os_type: String,
    pub os_version: String,
    pub app_version: String,
    pub available_sensors: Vec<String>,
    pub capabilities: DeviceCapabilities,
    pub created_at: DateTime<Utc>,
    pub last_seen: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DeviceCapabilities {
    pub has_camera: bool,
    pub has_microphone: bool,
    pub has_gps: bool,
    pub has_accelerometer: bool,
    pub has_gyroscope: bool,
    pub has_magnetometer: bool,
    pub has_proximity: bool,
    pub has_light: bool,
    pub has_pressure: bool,
    pub has_temperature: bool,
    pub has_humidity: bool,
    pub has_step_counter: bool,
    pub has_heart_rate: bool,
    pub has_ecg: bool,
    pub has_blood_oxygen: bool,
    pub has_stress: bool,
    pub has_compass: bool,
    pub screen_details: ScreenDetails,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ScreenDetails {
    pub width: i32,
    pub height: i32,
    pub density: f32,
    pub refresh_rate: i32,
}

impl Device {
    pub fn from_json(json: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(json)
    }

    pub fn to_json(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(self)
    }
}

impl DeviceCapabilities {
    pub fn from_json(json: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(json)
    }

    pub fn to_json(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(self)
    }
} 